src/app/shared/services/piece-layer.service.ts

```diff
@@ -0,0 +1,58 @@
+import { Piece } from '../../core/models/piece'
+import {Part} from "../../core/models/part";
+
+export class PieceLayerService {
+    public getPieceLayersForParts(parts: Part[]): string[] {
+        const pieceLayers: string[] = []
+        parts.forEach(part => {
+            part.pieces.concat(part.adLibPieces).forEach(piece => {
+                const layer = this.getPieceLayer(piece)
+                if (!pieceLayers.includes(layer) && layer !== 'UNKNOWN') {
+                    pieceLayers.push(layer)
+                }
+            })
+        })
```

Comment:
This isn't tested but this should return the same without the nested `foreach` loop:
```
const duplicatedLayers: string[] = parts.flatMap(part => {
            return part.pieces.concat(part.adLibPieces)
                .map(piece => this.getPieceLayer(piece))
                .filter(layer => layer !== 'UNKNOWN')
        })
const set: Set<string> = new Set(duplicatedLayers)
return Array.from(set)
```
We should try to limit the use of `foreach` and use some of the other methods available on Array instead. It's a big flag that the `foreach` are nested.

If you always want the returned array to contain unique entries, then consider just changing the return type to a Set.

Reply:
This has been rewritten to use a functional approach:
```typescript
    public getPieceLayersForParts(parts: Part[]): PieceLayer[] {
        const pieces: Piece[] = parts.flatMap(part => part.pieces.concat(part.adLibPieces))
        const pieceLayers: PieceLayer[] = pieces.map(piece => this.getPieceLayer(piece))
        return [...new Set(pieceLayers)]
    }
```

Reply:
Followup question: 
Why not let the return type be `Set<PieceLayer>` instead of `PieceLayer[]` if it's a requirement that there are no duplicates?

Reply:
Now returns `Set<PieceLayer>`.